go/kbfs/simplefs/archive.go

```diff
@@ -0,0 +1,876 @@
+	}
+}
+
+func (m *archiveManager) copyFile(ctx context.Context,
+	srcDirFS billy.Filesystem, entryPathWithinJob string,
+	localPath string, srcSeekOffset int64, mode os.FileMode) (sha256Sum []byte, err error) {
+	if srcSeekOffset == 0 {
+		return m.copyFileFromBeginning(ctx, srcDirFS, entryPathWithinJob, localPath, mode)
+	}
+	return m.copyFilePickupPrevious(ctx, srcDirFS, entryPathWithinJob, localPath, srcSeekOffset, mode)
+}
+
+func getWorkspaceDir(jobDesc keybase1.SimpleFSArchiveJobDesc) string {
+	return filepath.Join(jobDesc.StagingPath, "workspace")
+}
+
+func (m *archiveManager) doCopying(ctx context.Context, jobID string) (err error) {
+	m.simpleFS.log.CDebugf(ctx, "+ doCopying %s", jobID)
+	defer func() { m.simpleFS.log.CDebugf(ctx, "- doCopying %s err: %v", jobID, err) }()
+
+	desc, manifest := func() (keybase1.SimpleFSArchiveJobDesc, map[string]keybase1.SimpleFSArchiveFile) {
+		m.mu.Lock()
+		defer m.mu.Unlock()
+		manifest := make(map[string]keybase1.SimpleFSArchiveFile)
+		for k, v := range m.state.Jobs[jobID].Manifest {
+			manifest[k] = v.DeepCopy()
+		}
+		return m.state.Jobs[jobID].Desc, manifest
+	}()
+
+	updateManifest := func(manifest map[string]keybase1.SimpleFSArchiveFile) {
+		m.mu.Lock()
+		defer m.mu.Unlock()
+		// Can override directly since only one worker can work on a give job at a time.
+		job := m.state.Jobs[jobID]
+		for k, v := range manifest {
+			job.Manifest[k] = v.DeepCopy()
+		}
+		m.state.Jobs[jobID] = job
+	}
+
+	srcContainingDirFS, finalElem, err := m.simpleFS.getFSIfExists(ctx,
+		keybase1.NewPathWithKbfsArchived(desc.KbfsPathWithRevision))
+	if err != nil {
+		return fmt.Errorf("getFSIfExists error: %v", err)
+	}
+	srcDirFS, err := srcContainingDirFS.Chroot(finalElem)
+	if err != nil {
+		return fmt.Errorf("srcContainingDirFS.Chroot error: %v", err)
+	}
+	dstBase := filepath.Join(getWorkspaceDir(desc), desc.TargetName)
+
+	entryPaths := make([]string, 0, len(manifest))
+	for entryPathWithinJob := range manifest {
+		entryPaths = append(entryPaths, entryPathWithinJob)
+	}
+	sort.Strings(entryPaths)
+
+	for _, entryPathWithinJob := range entryPaths {
+		entry := manifest[entryPathWithinJob]
+		entry.State = keybase1.SimpleFSFileArchiveState_InProgress
+		manifest[entryPathWithinJob] = entry
+		updateManifest(manifest)
+
+		localPath := filepath.Join(dstBase, entryPathWithinJob)
+		srcFI, err := srcDirFS.Stat(entryPathWithinJob)
+		if err != nil {
+			return fmt.Errorf("srcDirFS.Stat(%s) error: %v", entryPathWithinJob, err)
+		}
+		switch {
+		case srcFI.IsDir():
+			os.MkdirAll(localPath, 0755)
+			if err != nil {
+				return fmt.Errorf("os.MkdirAll(%s) error: %v", localPath, err)
+			}
+			os.Chtimes(localPath, time.Time{}, srcFI.ModTime())
+			if err != nil {
+				return fmt.Errorf("os.Chtimes(%s) error: %v", localPath, err)
+			}
+			entry.State = keybase1.SimpleFSFileArchiveState_Complete
+			manifest[entryPathWithinJob] = entry
+		case srcFI.Mode()&os.ModeSymlink != 0: // symlink
+			panic("todo")
```

Comment:
What's the plan here?  A separate MR?  If so, would it be done before this is released?

Reply:
The embarrassing reality is I simply forgot. Will add something here

Reply:
hmm it looks like zip (or at least go's zip) doesn't support symlinks. Maybe we should switch to `.tar.gz`. I'll continue tomorrow

Reply:
I got symlink working for zip. It seems the reason why Go’s `zip` doesn't supoort symlinks in `AddFS` is because `io/fs` doesn’t support symlinks yet. So I changed our copy of the function to take a dir path rather than a `fs.FileSystem`